File: autocontext/src/autocontext/harness/core/controller.py
"""Domain-agnostic loop controller for pause/resume, gate override, hints, chat."""

from __future__ import annotations

import queue
import threading


class LoopController:
    """Thread-safe control interface for the generation loop."""

    def __init__(self) -> None:
        self._pause_event = threading.Event()
        self._pause_event.set()  # starts running (not paused)
        self._lock = threading.Lock()
        self._gate_override: str | None = None
        self._pending_hint: str | None = None
        self._pending_chat: queue.Queue[tuple[str, str]] = queue.Queue()
        self._chat_responses: queue.Queue[tuple[str, str]] = queue.Queue()

    def pause(self) -> None:
        self._pause_event.clear()

    def resume(self) -> None:
        self._pause_event.set()

    def is_paused(self) -> bool:
        return not self._pause_event.is_set()

    def wait_if_paused(self) -> None:
        """Block the calling thread until resumed."""
        self._pause_event.wait()

    def set_gate_override(self, decision: str) -> None:
        with self._lock:
            self._gate_override = decision

    def take_gate_override(self) -> str | None:
        with self._lock:
            val = self._gate_override
            self._gate_override = None
            return val

    def inject_hint(self, text: str) -> None:
        with self._lock:
            self._pending_hint = text

    def take_hint(self) -> str | None:
        with self._lock:
            val = self._pending_hint
            self._pending_hint = None
            return val

    def submit_chat(self, role: str, message: str) -> str:
        """Submit a chat request and block until the loop thread responds."""
        self._pending_chat.put((role, message))
        _role, response = self._chat_responses.get()
        return response

    def poll_chat(self) -> tuple[str, str] | None:
        """Non-blocking check for pending chat requests."""
        try:
            return self._pending_chat.get_nowait()
        except queue.Empty:
            return None

    def respond_chat(self, role: str, response: str) -> None:
        self._chat_responses.put((role, response))

File: ts/src/loop/controller.ts
/**
 * Loop controller — pause/resume state machine with Promise-based blocking (AC-342).
 * Mirrors Python's autocontext/harness/core/controller.py.
 */

export class LoopController {
  #paused = false;
  #resumeResolvers: Array<() => void> = [];
  #gateOverride: string | null = null;
  #pendingHint: string | null = null;
  #chatQueue: Array<{ role: string; message: string; resolve: (response: string) => void }> = [];
  #pendingChatResolvers: Array<(response: string) => void> = [];

  pause(): void {
    this.#paused = true;
  }

  resume(): void {
    this.#paused = false;
    // Resolve all waiting promises
    const resolvers = this.#resumeResolvers.splice(0);
    for (const resolve of resolvers) {
      resolve();
    }
  }

  isPaused(): boolean {
    return this.#paused;
  }

  waitIfPaused(): Promise<void> {
    if (!this.#paused) return Promise.resolve();
    return new Promise<void>((resolve) => {
      this.#resumeResolvers.push(resolve);
    });
  }

  setGateOverride(decision: string): void {
    this.#gateOverride = decision;
  }

  takeGateOverride(): string | null {
    const val = this.#gateOverride;
    this.#gateOverride = null;
    return val;
  }

  injectHint(text: string): void {
    this.#pendingHint = text;
  }

  takeHint(): string | null {
    const val = this.#pendingHint;
    this.#pendingHint = null;
    return val;
  }

  submitChat(role: string, message: string): Promise<string> {
    return new Promise<string>((resolve) => {
      this.#chatQueue.push({ role, message, resolve });
    });
  }

  pollChat(): [string, string] | null {
    if (this.#chatQueue.length === 0) return null;
    const entry = this.#chatQueue.shift()!;
    this.#pendingChatResolvers.push(entry.resolve);
    return [entry.role, entry.message];
  }

  respondChat(_role: string, response: string): void {
    if (this.#pendingChatResolvers.length > 0) {
      const resolve = this.#pendingChatResolvers.shift()!;
      resolve(response);
    }
  }
}
